parsl/data_provider/data_manager.py

```diff
@@ -110,39 +127,102 @@ def _get_globus_site(self, site_name=None):
                         'working_dir': working_dir}
         raise Exception('No site with a Globus endpoint and working_dir defined')
 
-    def stage_in(self, file, site_name=None):
+    def _get_working_dir(self, site_name):
+        for s in self.config['sites']:
+            if site_name == site_name:
+                if 'data' not in s:
+                    break
+                data = s['data']
+                if 'working_dir' not in data:
+                    break
+                return os.path.normpath(data['working_dir'])
+        return None
+
+    def stage_in(self, file, site_name):
         """Transport the file from the site of origin to the site.
 
         This function returns a DataFuture.
 
         Args:
             - self
-            - file (File) - file to stage in
-            - site_name (str) - a name of a site the file is going to be staged in to.
+            - file (File) : file to stage in
+            - site_name (str|List) : a site the file is going to be staged in to.
                                 If the site argument is not specified for a file
                                 with 'globus' scheme, the file will be staged in to
                                 the first site with the "globus" key in a config.
         """
 
         if file.scheme == 'file':
-            site_name = None
+            stage_in_app = self._file_stage_in_app()
+            app_fut = stage_in_app(outputs=[file])
+            return app_fut._outputs[0]
+        elif file.scheme == 'ftp':
+            working_dir = self._get_working_dir(site_name)
+            stage_in_app = self._ftp_stage_in_app(site_name=site_name)
+            app_fut = stage_in_app(working_dir, outputs=[file])
+            return app_fut._outputs[0]
+        elif file.scheme == 'http' or file.scheme == 'https':
+            working_dir = self._get_working_dir(site_name)
+            stage_in_app = self._http_stage_in_app(site_name=site_name)
+            app_fut = stage_in_app(working_dir, outputs=[file])
+            return app_fut._outputs[0]
         elif file.scheme == 'globus':
             globus_ep = self._get_globus_site(site_name)
+            stage_in_app = self._globus_stage_in_app()
+            app_fut = stage_in_app(globus_ep, outputs=[file])
+            return app_fut._outputs[0]
 
-        df = file.get_data_future(globus_ep['site_name'])
-        if df:
-            return df
-
-        if file.scheme == 'file':
-            f = self.submit(self._file_transfer_in, file)
-        elif file.scheme == 'globus':
-            f = self.submit(self._globus_transfer_in, file, globus_ep)
+    def _file_stage_in_app(self):
+        return App("python", self.dfk, sites=['data_manager'])(self._file_stage_in)
 
-        from parsl.app.futures import DataFuture
+    def _file_stage_in(self, outputs=[]):
```

Comment:
I don’t follow why `_file_stage_in/_file_stage_out` are no ops. Say, for example, you’re running the same app on two sites, Site A and Site B. You want results to be sent to `/home/foobar` on Site A. If you’re running on Site B, you use globus transfer, if you’re running on site A, you just want to use `mv`.  Is this possible? I think a test that has two sites, one with (for example) globus transfer, another with local transfer, and a task that runs on both sites would clarify this for me.

Reply:
I don't think I am doing a good job of communicating my concern. Let's take a few minutes to talk about this at the office tomorrow morning and then I'll follow up with a summary here.

Reply:
Talking with @yadudoc cleared this up for me. The answer to my question ('What happens if you are running on two sites?') is that multisite + data staging is not supported. @lukaszlacinski We need to add a warning that complains if the user tries to run a config with multiple sites + data management. I will open a separate issue to do that later and assign myself, because it should be done in the constructor of the new `Config` class.

Reply:
@annawoodard Multisite is supported. A remote input file is staged in on a site where an app that needs the file is going to be launched.

Reply:
We are better off advertising only single remote site support. I don't think implicit staging has been fleshed out to support data movement across multiple sites. For eg. AppB on SiteB consumes files from AppA on SiteA. It is unclear in this model if we'll stage all output data to local and then stage out.